tests/tensorflow/test_tensorflow_estimator_export.py

```diff
@@ -0,0 +1,273 @@
+            f"python {exec_py_path} {task_type} {save_as_type} "
+            f"{save_path if save_path else ''}",
+            install_mlflow=False,
+            command_env={"MLFLOW_TRACKING_URI": tracking_uri, "MLFLOW_REPO_PATH": mlflow_repo_path},
+        )
+        with open(output_data_file_path, "rb") as f:
+            inference_df, expected_results_df, run_id = pickle.load(f)
+            return (
+                ModelDataInfo(
+                    inference_df=inference_df,
+                    expected_results_df=expected_results_df,
+                    raw_results=None,
+                    raw_df=None,
+                    run_id=run_id,
+                ),
+                tracking_uri,
+            )
+
+
+def test_load_model_from_remote_uri_succeeds(tmpdir, model_path, mock_s3_bucket):
+    model_data_info, _ = save_or_log_tf_model_by_mlflow128(
+        str(tmpdir), "iris", "save_model", model_path
+    )
+
+    artifact_root = "s3://{bucket_name}".format(bucket_name=mock_s3_bucket)
+    artifact_path = "model"
+    artifact_repo = S3ArtifactRepository(artifact_root)
+    artifact_repo.log_artifacts(model_path, artifact_path=artifact_path)
+
+    model_uri = artifact_root + "/" + artifact_path
+    infer = mlflow.tensorflow.load_model(model_uri=model_uri)
+    feed_dict = {
+        df_column_name: tf.constant(model_data_info.inference_df[df_column_name])
+        for df_column_name in list(model_data_info.inference_df)
+    }
+    raw_preds = infer(**feed_dict)
+    pred_dict = {column_name: raw_preds[column_name].numpy() for column_name in raw_preds.keys()}
+    for col in pred_dict:
+        np.testing.assert_allclose(
+            np.array(pred_dict[col], dtype=np.float),
+            np.array(model_data_info.raw_results[col], dtype=np.float),
+        )
+
+
+def test_iris_model_can_be_loaded_and_evaluated_successfully(tmpdir, model_path):
+    model_data_info, _ = save_or_log_tf_model_by_mlflow128(
+        str(tmpdir), "iris", "save_model", model_path
+    )
+
+    def load_and_evaluate():
+
+        infer = mlflow.tensorflow.load_model(model_uri=model_path)
+        feed_dict = {
+            df_column_name: tf.constant(model_data_info.inference_df[df_column_name])
+            for df_column_name in list(model_data_info.inference_df)
+        }
+        raw_preds = infer(**feed_dict)
+        pred_dict = {
+            column_name: raw_preds[column_name].numpy() for column_name in raw_preds.keys()
+        }
+        for col in pred_dict:
+            np.testing.assert_array_equal(pred_dict[col], model_data_info.raw_results[col])
+
+    load_and_evaluate()
+
+    with tf.device("/CPU:0"):
+        load_and_evaluate()
+
+
+def test_log_and_load_model_persists_and_restores_model_successfully(tmpdir):
+    model_data_info, tracking_uri = save_or_log_tf_model_by_mlflow128(
+        str(tmpdir), "iris", "log_model"
+    )
+    model_uri = f"runs:/{model_data_info.run_id}/model"
+    with _use_tracking_uri(tracking_uri):
```

Comment:
Why do we need to call `_use_tracking_uri` here?

Reply:
The child process I use a file path tracking_uri.
In the parent pytest process, we need to read runs logged by the child process, so I use the same tracking uri used in child process.

Reply:
Why do we need to use a file tracking URI?

Reply:
> Can we pass the tracking URI from parent process to child process?

Yes. This is what current code does.
See `save_or_log_keras_model_by_mlflow128` helper function, it first generates a tracking_uri, then pass it to child process.

Reply:
Because the child process changed "working directory", if reusing parent process tracking URI, child process logged artifacts cannot be read by parent process because of relative mlruns path issue.

Reply:
Have you tried creating a new experiment with an absolute artifact URI?

Reply:
Feel free to ignore this suggestion if it complicates the code.

Reply:
It need more code, I think setting one local file tracking URI is simpler.

Reply:
I found a workaround:

```diff
diff --git a/tests/tensorflow/save_keras_model.py b/tests/tensorflow/save_keras_model.py
index 2b7a2885e..a73310e89 100644
--- a/tests/tensorflow/save_keras_model.py
+++ b/tests/tensorflow/save_keras_model.py
@@ -1,6 +1,5 @@
 import tensorflow as tf
 import sys
-import pandas as pd
 import numpy as np
 import pickle
 
@@ -15,7 +14,19 @@ model = tf.keras.Model(inputs=inputs, outputs=[outputs])
 
 task_type = sys.argv[1]
 save_as_type = sys.argv[2]
+save_path = sys.argv[3]
+tracking_uri = sys.argv[4]
+repo_path = sys.argv[5]
 
+# "python",
+# exec_py_path,
+# task_type,
+# save_as_type,
+# save_path if save_path else "",
+# mlflow.get_tracking_uri(),
+# mlflow_repo_path,
+
+mlflow.set_tracking_uri(tracking_uri)
 run_id = None
 
 if save_as_type == "tf1-estimator":
diff --git a/tests/tensorflow/test_keras_model_export.py b/tests/tensorflow/test_keras_model_export.py
index c597e3701..e9bba0519 100644
--- a/tests/tensorflow/test_keras_model_export.py
+++ b/tests/tensorflow/test_keras_model_export.py
@@ -595,10 +595,10 @@ def test_load_without_save_format(tf_keras_model, model_path):
     assert tf_keras_model.to_json() == model_loaded.to_json()
 
 
-@pytest.mark.skipif(
-    not (_is_importable("transformers") and keras_version >= Version("2.6.0")),
-    reason="This test requires transformers, which is no longer compatible with Keras < 2.6.0",
-)
+# @pytest.mark.skipif(
+#     not (_is_importable("transformers") and keras_version >= Version("2.6.0")),
+#     reason="This test requires transformers, which is no longer compatible with Keras < 2.6.0",
+# )
 def test_pyfunc_serve_and_score_transformers():
     from transformers import BertConfig, TFBertModel  # pylint: disable=import-error
 
@@ -650,18 +650,19 @@ def test_virtualenv_subfield_points_to_correct_path(model, model_path):
     assert python_env_path.is_file()
 
 
-def test_load_and_predict_keras_model_saved_by_mlflow128(tmpdir):
-    model_data_info, tracking_uri = save_or_log_keras_model_by_mlflow128(
-        str(tmpdir), task_type="log_model", save_as_type="keras"
+def test_load_and_predict_keras_model_saved_by_mlflow128(tmp_path, monkeypatch):
+    # Change the working directory to the temporary directory to ensure we use mlflow installed
+    # from PyPI
+    monkeypatch.chdir(tmp_path)
+    model_data_info = save_or_log_keras_model_by_mlflow128(
+        tmp_path, task_type="log_model", save_as_type="keras"
     )
-
     model_uri = f"runs:/{model_data_info.run_id}/model"
 
-    def load_and_predict(load_model_fn):
-        with _use_tracking_uri(tracking_uri):
-            mlflow_model = load_model_fn()
-        predictions = mlflow_model.predict(model_data_info.inference_df)
-        np.testing.assert_allclose(predictions, model_data_info.expected_results_df)
+    model = mlflow.pyfunc.load_model(model_uri)
+    predictions = model.predict(model_data_info.inference_df)
+    np.testing.assert_allclose(predictions, model_data_info.expected_results_df)
 
-    load_and_predict(lambda: mlflow.pyfunc.load_model(model_uri))
-    load_and_predict(lambda: mlflow.tensorflow.load_model(model_uri))
+    model = mlflow.tensorflow.load_model(model_uri)
+    predictions = model.predict(model_data_info.inference_df)
+    np.testing.assert_allclose(predictions, model_data_info.expected_results_df)
diff --git a/tests/tensorflow/test_tensorflow_estimator_export.py b/tests/tensorflow/test_tensorflow_estimator_export.py
index a850ac460..2d6e2e500 100644
--- a/tests/tensorflow/test_tensorflow_estimator_export.py
+++ b/tests/tensorflow/test_tensorflow_estimator_export.py
@@ -1,7 +1,9 @@
 # pep8: disable=E501
 
+import subprocess
 import collections
 import os
+from pathlib import Path
 import pickle
 import pytest
 import json
@@ -71,35 +73,38 @@ def save_or_log_tf_model_by_mlflow128(tmpdir, model_type, task_type, model_path=
             return ModelDataInfo(*pickle.load(f)), tracking_uri
 
 
-def save_or_log_keras_model_by_mlflow128(tmpdir, task_type, save_as_type, save_path=None):
-    conda_env = get_or_create_conda_env("tests/tensorflow/mlflow-128-tf-23-env.yaml")
-    output_data_file_path = os.path.join(tmpdir, "output_data.pkl")
-    tracking_uri = "file:" + os.path.abspath(os.path.join(tmpdir, "mlruns"))
-    exec_py_path = os.path.abspath("tests/tensorflow/save_keras_model.py")
-    mlflow_repo_path = os.getcwd()
-
-    with chdir(tmpdir):
-        # change current working directory to be a temporary directory,
-        # to prevent importing current repo mlflow module.
-        _execute_in_conda_env(
-            conda_env,
-            f"python {exec_py_path} {task_type} {save_as_type} "
-            f"{save_path if save_path else ''}",
-            install_mlflow=False,
-            command_env={"MLFLOW_TRACKING_URI": tracking_uri, "MLFLOW_REPO_PATH": mlflow_repo_path},
-        )
-        with open(output_data_file_path, "rb") as f:
-            inference_df, expected_results_df, run_id = pickle.load(f)
-            return (
-                ModelDataInfo(
-                    inference_df=inference_df,
-                    expected_results_df=expected_results_df,
-                    raw_results=None,
-                    raw_df=None,
-                    run_id=run_id,
-                ),
-                tracking_uri,
+def save_or_log_keras_model_by_mlflow128(tmp_path, task_type, save_as_type, save_path=None):
+    this_file_dir = Path(__file__).parent
+    conda_env = get_or_create_conda_env(this_file_dir / "mlflow-128-tf-23-env.yaml")
+    exec_py_path = os.path.join(this_file_dir / "save_keras_model.py")
+    repo_path = Path(mlflow.__file__).parent.parent
+    _execute_in_conda_env(
+        conda_env,
+        " ".join(
+            map(
+                str,
+                [
+                    "python",
+                    exec_py_path,
+                    task_type,
+                    save_as_type,
+                    save_path if save_path else "test",
+                    mlflow.get_tracking_uri(),
+                    repo_path,
+                ],
             )
+        ),
+        install_mlflow=False,
+    )
+    with open(tmp_path / "output_data.pkl", "rb") as f:
+        inference_df, expected_results_df, run_id = pickle.load(f)
+        return ModelDataInfo(
+            inference_df=inference_df,
+            expected_results_df=expected_results_df,
+            raw_results=None,
+            raw_df=None,
+            run_id=run_id,
+        )
 
 
 def test_load_model_from_remote_uri_succeeds(tmpdir, model_path, mock_s3_bucket):
@@ -138,9 +143,7 @@ def test_iris_model_can_be_loaded_and_evaluated_successfully(tmpdir, model_path)
         for df_column_name in list(model_data_info.inference_df)
     }
     raw_preds = infer(**feed_dict)
-    pred_dict = {
-        column_name: raw_preds[column_name].numpy() for column_name in raw_preds.keys()
-    }
+    pred_dict = {column_name: raw_preds[column_name].numpy() for column_name in raw_preds.keys()}
     for col in pred_dict:
         np.testing.assert_array_equal(pred_dict[col], model_data_info.raw_results[col])
 

```